Python Dispatcher integration with C++ dispatcher

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84966
* #84965
* __->__ #84826

Signed-off-by: Edward Z. Yang <ezyangfb.com>

From @ezyang's original PR:

There are a number of situations where we have non-backend kernels (e.g., CompositeImplicitAutograd, batching rules) which we would like to port to Python, but we have no way to integrate these ports with the overall system while using preexisting C++ registrations otherwise. This PR changes that by introducing a Python dispatcher (which can have its own kernels directly in Python), which can be interpose over ordinary C++ dispatch. The ingredients:

We introduce a new PythonDispatcher dispatch key, that has the same tenor as FuncTorchDynamicLayerFrontMode: it works by getting triggered before every other dispatch key in the dispatch key, and shunting to a Python implementation
The Python dispatcher is a per-interpreter global object that is enabled/disabled via the guard EnablePythonDispatcher/DisablePythonDispatcher. We don't make it compositional as I have no idea what a compositional version of this feature would look like. Because it is global, we don't need to memory manage it and so I use a simpler SafePyHandle (newly added) to control access to this pointer from non-Python C++. Like __torch_dispatch__, we use PyInterpreter to get to the Python interpreter to handle the dispatch.
I need to reimplement dispatch table computation logic in Python. To do this, I expose a lot more helper functions for doing computations on alias dispatch keys and similar. I also improve the pybind11 handling for DispatchKey so that you can either accept the pybind11 bound enum or a string; this simplifies our binding code. See https://github.com/pybind/pybind11/issues/483#issuecomment-1237418106 for how this works; the technique is generally useful.

I need to be able to call backend fallbacks. I do this by permitting you to call at a dispatch key which doesn't have a kernel for the operator; if the kernel doesn't exist, we check the backend fallback table instead.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

voznesenskym commented:
Fix after my git/ghstack mistake in #84551

voznesenskym commented:
Fix after my git/ghstack mistake in #84551

voznesenskym commented on functorch/functorch/experimental/cond.py:
```diff
@@ -117,21 +123,17 @@ def cond_autograd(pred, true_fn, false_fn, *operands):
-
-    return inner
+@cond.py_impl(DispatchKey.Python)
+@cond.py_impl(ProxyTorchDispatchMode)
```
@ezyang thoughts on explicit mode registration? the idea is that if we see dispatch for an op, we assert that the given op is registered for the given mode. 

Should this be exclusion? Also we do not dispatch on modes yet, only keys, so we must assert that the impls match. Should we ever relax this assertion in the future?

voznesenskym replied:
Actually, I have a better idea here...

ezyang commented on functorch/functorch/experimental/cond.py:
```diff
@@ -117,21 +123,16 @@ def cond_autograd(pred, true_fn, false_fn, *operands):
+# TODO(voz): Make this automatic for keys, this is very ugly atm
+cond.fallthrough(DispatchKey.PythonTLSSnapshot)
+cond.fallthrough(DispatchKey.ADInplaceOrView)
+cond.fallthrough(DispatchKey.BackendSelect)
```
So, re fallthrough, I think the easiest fix will be to add C bindings to query the set of keys in c++ which are fallthrough, and then override it as necessary based on python registrations. cond.fallthrough is not symmetric with how fallthrough registrations work in C++, which is that fallthrough is a special "function" you can register.

voznesenskym replied:
Agreed, this one is important enough to leave an issue behind, let me do that rq

ezyang commented on torch/_dispatch/python.py:
```diff
@@ -0,0 +1,86 @@
+    source = f'torch.ops.{op}.dispatch(k, *args, **kwargs)'
+    filename = f'{op}[{torch._C._dispatch_key_name(k)}]'
+    compiled = compile(source, filename, 'eval')  # TODO: maybe cache?
+    return eval(compiled, {'torch': torch, 'k': k, 'args': args, 'kwargs': kwargs})
```
This can be improved a lot, but this is an OK start

voznesenskym replied:
Let me add a comment about caching this and potentially not recreating the objects each time...

ezyang commented on torch/_ops.py:
```diff
@@ -26,9 +32,127 @@ def dl_open_guard():
+        self.python_key_mode_table = {}
+
+        # Make _OPNamespace not scream, this whole name based association needs a good hard look
+        self.__name__ = "pyop." + name
```
TBH, there isn't really any good reason to make pyops available at torch.ops.pyop. It is necessary for c++ bound ops as we must introduce them to Python namespace somewhere, but for Python ops you can use the normal namespacing mechanisms of Python to handle their namespacing

voznesenskym replied:
This is to hack away at the first getattr. Removing this "pyop." addition gives us:

```
TypeError: '_OpNamespace' object is not callable
```
Hence the vague comment. Its a bit of a huge hack to get around this:

```
    def __getattr__(self, name):
        # Here we are creating `torch.ops.my_namespace`
        namespace = _OpNamespace(name)
        setattr(self, name, namespace)
        return namespace
```

Alternatively, we could keep a global list to check, and if the op is there, skip the namespace notion, or have a "default" namespace of sorts we consult first...

ezyang commented on torch/_ops.py:
```diff
@@ -26,9 +32,127 @@ def dl_open_guard():
+
+    def py_impl(self, dispatch_key_or_mode):
+        def inner(fn):
+            if isinstance(dispatch_key_or_mode, type(TorchDispatchMode)):
```
don't you want issubclassof

voznesenskym replied:
Sure, that's cleaner

ezyang replied:
It's not just cleaner, the current code is wrong lol (in the absence of metaclass the type will always be type)

ezyang commented on torch/_ops.py:
```diff
@@ -26,9 +32,127 @@ def dl_open_guard():
+                curr_mode in self.python_key_mode_table
+            ), f"Current active mode {curr_mode} not registered"
+            # TODO(voz): The idea behind this is that we do not yet support dispatch by key + mode, only key.
+            return self.python_key_mode_table[curr_mode](*args, **kwargs)
```
Some (most?) mode implementations will need their mode object `curr_mode` as it contains essential state for doing the mode op.

voznesenskym replied:
yea, but I was thinking they could query the global mode accessor - should we pass it to them?

ezyang replied:
when you call into the mode, that mode should already be removed from the stack. So yes we need to suspend on their behalf

ezyang commented on torch/csrc/utils/python_dispatch.cpp:
```diff
@@ -332,46 +341,66 @@ void initDispatchBindings(PyObject* module) {
+                  DEF_ONE(CompositeImplicitAutograd) DEF_ONE(AutogradOther)
+                      DEF_ONE(Autograd) DEF_ONE(BackendSelect)
+                          DEF_ONE(ADInplaceOrView) DEF_ONE(PythonTLSSnapshot)
+                              DEF_ONE(Python)
```
we should turn off clang format lol

voznesenskym commented:
pytorchbot  -m

voznesenskym commented:
@pytorchbot -m

voznesenskym commented:
@pytorchbot merge

ezyang commented on torch/_ops.py:
```diff
@@ -64,7 +64,8 @@ def fallthrough(self, dispatch_key):
         def inner(fn):
             if inspect.isclass(dispatch_key_or_mode) and issubclass(
-                dispatch_key_or_mode, TorchDispatchMode
+<<<<<<< HEAD
```
yarp

malfet commented:
@pytorchbot revert -m "Broke dynamo, see https://hud.pytorch.org/pytorch/pytorch/commit/35f6a69191ef762cf22b6cbfe94b8d9406e16674" -c  weird

voznesenskym commented:
@malfet what was pushed was not what was in the PR...

voznesenskym commented:
https://github.com/pytorch/pytorch/commit/35f6a69191ef762cf22b6cbfe94b8d9406e16674 differs from this state of the PR? Very strange.

voznesenskym commented:
I can confirm that as it is, this branch passes dynamo tests.

voznesenskym committed:
WIP: Python Dispatcher integration with C++ dispatcher

Signed-off-by: Edward Z. Yang <ezyangfb.com>

pick e219ce8a61

Integrate python-only dispatching, make pyop tests pass

Test fixes, mypy stuff

Lints and defines

small matmul change

Make matmul test happy

small test fix

fix tos

More test fixes

Save PythonDispatcherTLS to ThreadLocalState

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

WIP: Python Dispatcher integration with C++ dispatcher

Signed-off-by: Edward Z. Yang <ezyangfb.com>

pick e219ce8a61

Integrate python-only dispatching, make pyop tests pass

Test fixes, mypy stuff

Lints and defines

small matmul change

Make matmul test happy

small test fix

fix tos

More test fixes

Save PythonDispatcherTLS to ThreadLocalState

Signed-off-by: Edward Z. Yang <ezyangfb.com>

Make py_impl less error prone, make decompose use py_kernels

Wire up matmul and broadcast_tensors for real

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

fake file object for better backtrace

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Make _dispatch_key_name and parse more useful

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Reorder

Merge branch 'gh/voznesenskym/12/head' of github.com:pytorch/pytorch into gh/voznesenskym/12/head

Fix typo

expect pass

Mode registration

Better comment

Better mode registration

Fix incorrect assertion

purge matmul, fix string

Test removal

Lints, small fixes

fix attr mypy

pybind hack...

lint

better check

More linter stuff

Please work, linter

More linter woes

Update base for Update on "Python Dispatcher integration with C++ dispatcher"

Signed-off-by: Edward Z. Yang <ezyangfb.com>

From ezyang's original PR:

There are a number of situations where we have non-backend kernels (e.g., CompositeImplicitAutograd, batching rules) which we would like to port to Python, but we have no way to integrate these ports with the overall system while using preexisting C++ registrations otherwise. This PR changes that by introducing a Python dispatcher (which can have its own kernels directly in Python), which can be interpose over ordinary C++ dispatch. The ingredients:

We introduce a new PythonDispatcher dispatch key, that has the same tenor as FuncTorchDynamicLayerFrontMode: it works by getting triggered before every other dispatch key in the dispatch key, and shunting to a Python implementation
The Python dispatcher is a per-interpreter global object that is enabled/disabled via the guard EnablePythonDispatcher/DisablePythonDispatcher. We don't make it compositional as I have no idea what a compositional version of this feature would look like. Because it is global, we don't need to memory manage it and so I use a simpler SafePyHandle (newly added) to control access to this pointer from non-Python C++. Like __torch_dispatch__, we use PyInterpreter to get to the Python interpreter to handle the dispatch.
I need to reimplement dispatch table computation logic in Python. To do this, I expose a lot more helper functions for doing computations on alias dispatch keys and similar. I also improve the pybind11 handling for DispatchKey so that you can either accept the pybind11 bound enum or a string; this simplifies our binding code. See https://github.com/pybind/pybind11/issues/483#issuecomment-1237418106 for how this works; the technique is generally useful.

I need to be able to call backend fallbacks. I do this by permitting you to call at a dispatch key which doesn't have a kernel for the operator; if the kernel doesn't exist, we check the backend fallback table instead.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

WIP: Python Dispatcher integration with C++ dispatcher

Signed-off-by: Edward Z. Yang <ezyangfb.com>

pick e219ce8a61

Integrate python-only dispatching, make pyop tests pass

Test fixes, mypy stuff

Lints and defines

small matmul change

Make matmul test happy

small test fix

fix tos

More test fixes

Save PythonDispatcherTLS to ThreadLocalState

Signed-off-by: Edward Z. Yang <ezyangfb.com>

Ed fixes my linter woes

Cleanup

Merge branch 'gh/voznesenskym/12/head' of github.com:pytorch/pytorch into gh/voznesenskym/12/head

Fix clowny stuff

Fix dynamo failure by unhacking op names

WIP: Python Dispatcher integration with C++ dispatcher

Signed-off-by: Edward Z. Yang <ezyangfb.com>

pick e219ce8a61

Integrate python-only dispatching, make pyop tests pass

Test fixes, mypy stuff

Lints and defines

small matmul change

Make matmul test happy

small test fix

fix tos

More test fixes

Save PythonDispatcherTLS to ThreadLocalState

Signed-off-by: Edward Z. Yang <ezyangfb.com>

ghstack-source-id: 2fca1533c653eb29c0ae1d300a6de5041420af4d
Pull Request resolved: https://github.com/pytorch/pytorch/pull/84826

Ed fixes my linter woes

ghstack-source-id: 01aa659a09a607af44583ac3813ba39817d9b420
Pull Request resolved: https://github.com/pytorch/pytorch/pull/84965

Cleanup

ghstack-source-id: 29b0515b75590fed65e4a2763be1348454df09fe
Pull Request resolved: https://github.com/pytorch/pytorch/pull/84966

WIP: Python Dispatcher integration with C++ dispatcher

Signed-off-by: Edward Z. Yang <ezyangfb.com>

pick e219ce8a61

Integrate python-only dispatching, make pyop tests pass

Test fixes, mypy stuff

Lints and defines

small matmul change

Make matmul test happy

small test fix

fix tos

More test fixes

Save PythonDispatcherTLS to ThreadLocalState

Signed-off-by: Edward Z. Yang <ezyangfb.com>

ghstack-source-id: f7c212074fdd5fe7758b528acb1cab51d02e463e
Pull Request resolved: https://github.com/pytorch/pytorch/pull/84826

Make py_impl less error prone, make decompose use py_kernels

Wire up matmul and broadcast_tensors for real

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

fake file object for better backtrace

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Make _dispatch_key_name and parse more useful

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Reorder

WIP: Python Dispatcher integration with C++ dispatcher

Signed-off-by: Edward Z. Yang <ezyangfb.com>

pick e219ce8a61

Integrate python-only dispatching, make pyop tests pass

Test fixes, mypy stuff

Lints and defines

small matmul change

Make matmul test happy

small test fix

fix tos

More test fixes

Save PythonDispatcherTLS to ThreadLocalState

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

Fix typo

expect pass

Mode registration

Better comment

Better mode registration

Fix incorrect assertion

purge matmul, fix string

Test removal

Lints, small fixes

fix attr mypy

pybind hack...

lint

better check

More linter stuff

Please work, linter

More linter woes

Fix clowny stuff

Fix dynamo failure by unhacking op names

malfet commented:
> @malfet what was pushed was not what was in the PR...

Ok, this is 2nd change like that, so I need to write a linter into a mergebot. 
`ghstack` authored commits do not support hybrid dev mode: one can not author commit via `ghstack` and then just push changes to `head`

Will work on change ASAP that will prohibit merges to commits that do not have `ghstack-poisoned` beadcrumb in their description ( cc: @ezyang )

voznesenskym committed:
Fix more issues

ezyang commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -92,6 +92,7 @@ bool should_allow_numbers_as_tensors(const std::string& name) {
       "sub",          "sub_",          "sub_out",
       "subtract",     "subtract_",     "subtract_out", // alias of sub
       "true_divide",  "true_divide_",  "true_divide_out",
+      "to",           "_to_copy",      "copy_",
```
What's this for? This seems unrelated to this PR

voznesenskym replied:
This was to get certain tests passing

ezyang commented:
@malfet an alternate strategy which is more robust and IMO simpler to implement is to compare the trees of orig and head; if they are diverged, then refuse to merge.

voznesenskym commented:
Closed in https://github.com/pytorch/pytorch/pull/85050

bdhirsh commented on c10/core/DispatchKeySet.h:
```diff
@@ -172,7 +172,9 @@ class DispatchKeySet final {
                  1)) -
-            1) {}
+            1) {
+    *this = add(DispatchKey::PythonDispatcher);
```
Mostly as a call out (and reminder to myself - just spent a while trying to understand this line haha):

I think one consequence of this is that we need to be careful to use the right API's if we want to the PythonDispatcher to trigger properly during redispatches. AKA, this will properly trigger the PythonDispatcher when it's enabled:
```
// FULL_AFTER ctr will implicitly add the PythonDispatcher key to the keyset, so we can call into it
op.redispatchBoxed(ks & DispatchKeySet(DispatchKeySet::FULL_AFTER, my_dispatch_key));
```

but this redispatch will silently ignore the python dispatcher, even if it's enabled:
```
// PythonDispatcher key never gets added to the keyset, so even if we "turned it on", we'll ignore it
op.redispatchBoxed(DispatchKeySet({my_dispatch_key}));
```

I did a quick audit of `redispatch()` call-sites and I don't think there are an existing places where that happens today that are meaningful to tracing.
